Preview decision and remove individual keys

jmfrank63 opened the pull request:
Hi Andreas,

This might be a weird PR, but I have a use case for it. I am rate-limiting login failures, and thus would like to know in advance before I send a request to the rather costly authentication, what the decision would be, so in case of a failure I reject right away.
I considered wrapping first, but that would have meant to deal with the whole rate limiter for an individual decision.
Also, I wanted to have the ability to remove a key from the rate limiter, if some condition happens. The use case here is a user login. If it fails too often it gets rate limited, but if it succeeds it is reset.

I am a rather average coder, so I basically duplicated the code. But I am happy to take suggestions. Two things come to mind:
Call `check_key` for `peek_check_key` instead of duplicating the code. Use `fetch_and_update` instead of the `compare_exchange_weak` while loop.

I added more tests, as I intend to use this code in production. Please let me know what you think.

antifuchs commented:
I meant to do this, as well! Thanks so much for the PR (and sorry for not getting to reviewing it&the previous PR earlier - work combined with some sickness). I think having a `peek` method is great & promised to write one in some other issues (#131 at least!), so this is much appreciated (:

antifuchs commented on governor/src/gcra.rs:
```diff
@@ -111,6 +111,42 @@ impl Gcra {
+        let t0 = t0.duration_since(start);
+        let tau = self.tau;
+        let t = self.t;
+        match state.measure_and_peek(key, |tat| {
```
Hmmm, I had a feeling that it should be possible to use `measure_and_replace` by having the closure return `MW::disallow` and doing some manual math, in https://github.com/antifuchs/governor/compare/fix-120. This doesn't really work though, so duplicating that code and getting it right, there, seems better (:

jmfrank63 replied:
Sorry, for the delayed response, I ran stable formatting now.

antifuchs reviewed (approved):
This looks right! I'd like to get #144 in first, so that this looks a bit cleaner with the clippy fix-ups, but overall this is great!

jmfrank63 commented:
I rebased but since I fist had tried to resolve the conflicts via the web gui (a really bad idea) rebasing was a bit messy. I hope I didn't break anything, at least the test suite pased.

jmfrank63 commented:
@antifuchs Hi Andreas, I think I could need some help here. My tests are failing if no standard library is used. Could you please give me a hint on what to do?

jmfrank63 commented:
@antifuchs OK, I found the issue and fixed it. I added more tests to get the pipeline "green" as well as increasing code coverage slightly.
Since we are using the preview in our code, we currently have a fork hosted, which we have to keep in sync. Therefore, I have a strong interest to get back to an external reference of the crate.
Please let me know if there is anything else for me to do.

jmfrank63 commented:
@antifuchs Happy new year, Andreas. 
I just wanted to ask if there is anything else I could do. I don't seem to have any control over `bors`, so if you could have a look, that would be great.

Thanks, Johannes

antifuchs commented:
Sorry for taking so long with this. I'm coming into a good bit of spare time soon, so I'll review this in depth ASAP!

Nutomic commented:
I also need this functionality. Would be great if this could be merged soon, because there seems to be no other Rust crate with this functionality.

jmfrank63 commented:
Hi @antifuchs,

I was just revisiting this, it does not seem to need a rebase. Is there any other work you want me to do to get this merged?
I also saw, others seem to be interested in this, too.

Thanks, Johannes

antifuchs commented on governor/src/state/in_memory.rs:
```diff
@@ -44,6 +44,33 @@ impl InMemoryState {
+                }
+                Err(next_prev) => prev = next_prev,
+            }
+            decision = f(NonZeroU64::new(prev).map(|n| n.get().into()));
```
Looks like codecov thinks L65-L67 aren't exercised in tests; would be great if there was a way to test those too.

jmfrank63 replied:
I think I got this one now covered. Interestingly, now that `codecov` is succeeding, I cannot find the individual cover report any more. Should there be anything left, please let me know.

antifuchs commented on governor/src/state/in_memory.rs:
```diff
@@ -59,6 +86,16 @@ impl StateStore for InMemoryState {
+    }
+    fn reset(&self, _key: &Self::Key) {
+        self.0.store(0, Ordering::Release);
+    }
```
Same on coverage of the `reset` function - I think it'd be great if this could be called in tests at least (:

antifuchs commented on governor/src/state.rs:
```diff
@@ -46,6 +46,13 @@ pub trait StateStore {
     fn measure_and_replace<T, F, E>(&self, key: &Self::Key, f: F) -> Result<T, E>
     where
         F: Fn(Option<Nanos>) -> Result<(T, Nanos), E>;
+    /// `measure_and_peek` does the same as `measure_and_replace` except
```
formatting nit:
```suggestion

    /// `measure_and_peek` does the same as `measure_and_replace` except
```

jmfrank63 replied:
Correct, added the line.

antifuchs commented on governor/src/state/keyed.rs:
```diff
@@ -98,6 +98,15 @@ where
         )
     }
 
+    pub fn peek_key(&self, key: &K) -> Result<MW::PositiveOutcome, MW::NegativeOutcome> {
```
Please add a documentation comment on this public function.

jmfrank63 replied:
Docstring added. Docs are often a matter of taste. Please let me know if you want me to change anything.

antifuchs commented on governor/src/state/keyed.rs:
```diff
@@ -98,6 +98,15 @@ where
+            .peek_test::<K, C::Instant, S, MW>(self.start, key, &self.state, self.clock.now())
+    }
+
+    pub fn reset_key(&self, key: &K) {
```
public function: needs a doc comment

jmfrank63 replied:
Same here, docstring added.

antifuchs commented on governor/src/state/direct.rs:
```diff
@@ -80,6 +80,15 @@ where
         )
     }
 
+    pub fn peek(&self) -> Result<MW::PositiveOutcome, MW::NegativeOutcome> {
```
public function: needs a doc comment

antifuchs commented on governor/src/state/direct.rs:
```diff
@@ -107,6 +116,19 @@ where
             )
     }
+
+    pub fn peek_n(
```
public function: needs a doc comment

jmfrank63 replied:
Also done.

antifuchs commented on governor/tests/future.rs:
```diff
@@ -8,14 +8,13 @@ use std::sync::Arc;
-const MAX_TEST_RUN_DURATION: Duration = Duration::from_micros(200);
+const PROCEEDS_DURATION_MAX_MICROS: u64 = 250;
+const PAUSES_DURATION_MIN_MILLIS: u64 = 100;
+const MULTIPLE_DURATION_MIN_MILLIS: u64 = 8;
```
I believe the current main branch has a fix to these on it already (think it was https://github.com/antifuchs/governor/issues/157) - so I'm not sure if these extra constants are needed as workarounds to the flakey tests anymore.

jmfrank63 replied:
I removed them for now. The tests for time duration are actually pointless. All we need is to make sure the test succeeds. I have therefore removed all tests against times as this is really flaky. 
Please let me know if you don't like it, I will undo it.

antifuchs commented on governor/tests/keyed_dashmap.rs:
```diff
@@ -46,6 +46,215 @@ fn rejects_too_many() {
+
+    for key in KEYS {
+        // use up our burst capacity (2 in the first second):
+        assert!(
+            lb.check_key_n(key, nonzero!(2u32)).is_ok(),
+            "Now: {:?}",
+            clock.now()
+        );
```
codecov warns that lines in this multi-line assert aren't covered which... yeah ):

I think the easiest way to get rid of that is to reformat the condition you're checking in a let-bound variable and then assert on the variable.
```suggestion
        let res = lb.check_key_n(key, nonzero!(2u32));
        assert!(res.is_ok(), "Now: {:?}", clock.now());
```

(would be neat if we had `assert_matches` for these checks, but that's not stable yet)

jmfrank63 replied:
I think I got a solution. What is actually not covered here is the case the `check_n_key` fails.

antifuchs commented on governor/tests/keyed_dashmap.rs:
```diff
@@ -46,6 +46,215 @@ fn rejects_too_many() {
+            lb.check_key_n(key, nonzero!(2u32)).is_ok(),
+            "Now: {:?}",
+            clock.now()
+        );
```
ditto on coverage of that multiline-assert.

jmfrank63 replied:
Yeah, took a while, but I think I got it now. If I missed anything, please let me know.

antifuchs reviewed (commented):
This does look great! Besides a rebase, I highlighted a few items that still need work (mostly coverage-related, which will prevent a merge... but also missing docs on public methods).

I have owed you this review for a long time, thanks for picking up work on this change again!

jmfrank63 commented:
> This does look great! Besides a rebase, I highlighted a few items that still need work (mostly coverage-related, which will prevent a merge... but also missing docs on public methods).
> 
> I have owed you this review for a long time, thanks for picking up work on this change again!

@antifuchs Hi Andreas,

thank you so much for this opportunity. I think I have addressed all the issues. I also factored out two tests in their own functions and added one. For the collision, I changed the current test, it now succeeds really fast, I had successes with 4 retries. I have also removed the flaky timing tests on the proceed functions, all that is necessary is the fact that they proceed, not how fast.
If you don't like anything, just let me know, I'll change this.
If you know how to get the coverage report, like when it is failing, give me a hint. I couldn't find it :-)
Thanks again Johannes

jmfrank63 commented:
Hi Andreas,

Finally, full coverage, since I found the coverage report again. I also found that the M1 bug is now fixed, so I activated this test as well. I think you will decide on the version, so I left the version untouched.

jmfrank63 commented:
@antifuchs Hi Andreas,
I am not sure if I got all your concerns. The long time in between visiting this accumulated a lot of changes. I hope I have addressed all of them. While I am not at the company I used the code for actively in production, I care about my former colleagues. Currently, they have to maintain the fork. It would be nice if they could just return to a simple cargo dependency.
If I should do additional work on this, let me know.

jmfrank63 commented:
@antifuchs 
Hi Andreas,
I saw you are preparing for the next release. I would be happy if my PR could make it in there. I have increased my code coverage and resolved all issues I could find. All tests passed.
Please let me know if I should address anything else.
Thanks, Johannes

jmfrank63 commented:
@antifuchs Dear Andreas, Happy new year 2024.
I've rebased to reflect the latest changes, but I cannot do this forever. I am not invested into this any more as I have left the company, but for my colleagues it is painful to keep a local copy of an earlier version. I have added additional tests to increase coverage and have, to my best ability, addressed all concerns.
Therefore, I kindly ask you to consider merging now, so we can close this PR. You might want to review again, that's fine, I will try to address additional concerns ASAP.

Thanks, Johannes
